source/common/network/io_socket_handle_impl.cc

```diff
@@ -113,8 +113,8 @@ Api::IoCallUint64Result IoSocketHandleImpl::read(Buffer::Instance& buffer, uint6
     return Api::ioCallUint64ResultNoError();
   }
   Buffer::Reservation reservation = buffer.reserveApproximately(max_length);
-  Api::IoCallUint64Result result =
-      readv(reservation.length(), reservation.slices(), reservation.numSlices());
+  Api::IoCallUint64Result result = readv(std::min(reservation.length(), max_length),
```

Comment:
It's currently not, due to the heuristic in reserveApproximately. If the requested size is small, it ends up returning an entire slice. 

I have mixed feelings about that. Probably will result in less confusion to make it always <= max_length, but that may mean allocating some space that isn't available to the reservation.

Reply:
I think that reserveApproximately is only used to allocate relatively large amounts of memory.  I would consider ASSERT(preferred_length >= 16384); at the top of the OwnedImpl version of this method as a mitigation for this issue with small reservations.

Reply:
I thought so too, but a test failed (which is when I added `min` here), because this is the only `read()` call on a socket (roughly). There are tests that are dealing with a specific protocol and read the length of a specific header field, for instance, and don't expect extra data.